ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs

```diff
@@ -96,13 +137,19 @@ protocolInfoByron (NumCoreNodes numCoreNodes) nid params gc mLeader =
       , pInfoInitState  = ()
       }
   where
-    proofOfCredentials :: CoreNodeId -> PbftLeaderCredentials -> PBftIsLeader PBftCardanoCrypto
-    proofOfCredentials nid' (PbftLeaderCredentials sk dlg) =
+    delegation :: Cardano.Genesis.GenesisDelegation
+    delegation = Cardano.Genesis.configHeavyDelegation gc
+    proofOfCredentials :: Cardano.Genesis.GenesisDelegation -> PbftLeaderCredentials -> PBftIsLeader PBftCardanoCrypto
+    proofOfCredentials dlg cred@(PbftLeaderCredentials sk cert) =
       PBftIsLeader
-      { pbftCoreNodeId = nid' -- TODO (Duncan): compute this based on the genesis verification key
+      { pbftCoreNodeId = if checkDelegationSubject vk cert
+                         then flip fromMaybe (delegateNodeIdPbft dlg vk) $
+                              throw $ ByronInconsistentLeaderCredentials cred
+                         else throw $ ByronInconsistentLeaderCredentials cred
```

Comment:
In principle we should do this check in an earlier context, and not throw pure errors for what is a validation check, not a programmer error.

I'd suggest pulling this out as a top level exported verification function, and using it when resolving the config in the application. If you do that, you can still make it a check here, like an assertion, but then you have to document it as a pre-condition and provide that way of the app level doing the check first.

Here's a nice way to do it: don't export the `PbftLeaderCredentials` constructor but export a function that makes one, and it can do the check and return `Either ByronInconsistentLeaderCredentials PbftLeaderCredentials`. Though actually that adds zero extra information, so just use `Maybe PbftLeaderCredentials` and then you don't need `ByronInconsistentLeaderCredentials` at all.

Reply:
@dcoutts, this raises a philosophical question in my head -- if we're not exporting the dumb constructor from `ProtocolInfo.Byron`, are we ok with an expression that assumes the consistency ensured by the smart constructor?

Reply:
Yes, smart constructors that establish invariants is a standard pattern.